Stryker quits with "TypeError: Cannot read property 'runResult' of undefined"

I'm unable to run stryker on my current project. Error log and config below:

#### Error log

```
[2017-08-02 15:54:40.571] [WARN] InputFileResolver - Globbing expression "test/unit/**/*.js" did not result in any files.
[2017-08-02 15:54:40.595] [INFO] InputFileResolver - Found 33 of 33 file(s) to be mutated.
[2017-08-02 15:54:40.597] [INFO] SandboxCoordinator - Starting initial test run. This may take a while.
[2017-08-02 15:54:41.787] [ERROR] Stryker - One or more tests errored in the initial test run:
        TypeError: Cannot read property 'runResult' of undefined
TypeError: Cannot read property 'runResult' of undefined
```

#### stryker.conf.js

```
module.exports = function(config){
  config.set({
    files: [
      'test/unit/**/*.js',
      { pattern: 'src/**/*.js', included: false, mutated: true }
    ],
    testFramework: 'mocha',
    testRunner: 'mocha',
    reporter: ['progress', 'clear-text', 'dots', 'html', 'event-recorder'],
    coverageAnalysis: 'off',
    plugins: ['stryker-mocha-runner', 'stryker-html-reporter']
  });
};
```

Command I'm running to generate this error:

```
./node_modules/.bin/babel-node ./node_modules/.bin/stryker run
```

_Note: I've had a look in the `.stryker-tmp` directory, and the `___testHooksForStryker.js` file is empty. Assuming this is incorrect?_

The issues _I think_ I'm facing now are down to ES6 and bootstrapping. At the moment, when we run tests, we use `tests/bootstrap.js` which does quite a bit (examples below):

```
require('app-module-path').addPath(__dirname + '/../src/');
require('app-module-path').addPath(__dirname);

global.sinon = require('sinon');
```

Below are the errors I'm facing which I believe are related:

### tests/unit/lib/git/add_test.js

The error that halts stryker is as follows. The test that's causing this tries to import a module that _does_ exist, but relies on `bootstrap.js` to do some setup.

```
import {add} from 'lib/git';
```

Error generated by stryker.

```
[2017-08-03 09:48:25.515] [ERROR] Stryker - One or more tests errored in the initial test run:
        Error: MODULE_NOT_FOUND (undefined) Error: Cannot find module 'lib/git'
```

### src/config/data.js

I'm also seeing a lot of errors output in the console which _I believe_ are related to ES6 support. 

```
import fs from 'fs';
// ...
```

This generates the following error (I have quite a few of these, mostly moaning about line 1, which usually consists of `import`:

```
[2017-08-03 09:48:24.595] [ERROR] CoverageInstrumenterStream - Error while instrumenting file "/Users/lsteadman/FU/product-marketing-info-tool/src/config/data.js", error was: Error: Line 1: Unexpected token
Failed to parse file: /src/config/git.js
```

--- a/package.json
+++ b/package.json
@@ -14,7 +14,7 @@
     "@types/lodash": "^4.14.34",
     "@types/log4js": "0.0.32",
     "@types/mkdirp": "^0.3.28",
-    "@types/mocha": "^2.2.34",
+    "@types/mocha": "^2.2.42",
     "@types/mz": "0.0.31",
     "@types/progress": "^1.1.28",
     "@types/rimraf": "0.0.28",

--- a/packages/stryker-mocha-runner/src/MochaTestRunner.ts
+++ b/packages/stryker-mocha-runner/src/MochaTestRunner.ts
@@ -2,10 +2,7 @@ import * as log4js from 'log4js';
 import { EventEmitter } from 'events';
 import { TestRunner, RunResult, RunStatus, RunnerOptions } from 'stryker-api/test_runner';
 import { InputFile } from 'stryker-api/core';
-
-
-// import * as Mocha from 'mocha';
-const Mocha = require('mocha');
+import * as Mocha from 'mocha';
 import StrykerMochaReporter from './StrykerMochaReporter';
 
 const log = log4js.getLogger('MochaTestRunner');
@@ -25,12 +22,23 @@ export default class MochaTestRunner extends EventEmitter implements TestRunner 
     return new Promise<RunResult>((resolve, fail) => {
       try {
         this.purgeFiles();
-        let mocha = new Mocha({ reporter: StrykerMochaReporter, bail: true });
+        let mocha = new Mocha({ reporter: StrykerMochaReporter as any, bail: true });
         this.files.filter(file => file.included).forEach(f => mocha.addFile(f.path));
         try {
-          let runner: any = mocha.run((failures: number) => {
-            let result: RunResult = runner.runResult;
-            resolve(result);
+          mocha.run((failures: number) => {
+            const reporter = StrykerMochaReporter.CurrentInstance;
+            if (reporter) {
+              let result: RunResult = reporter.runResult;
+              resolve(result);
+            } else {
+              const errorMsg = 'The StrykerMochaReporter was not instantiated properly. Could not retrieve the RunResult.';
+              log.error(errorMsg);
+              resolve({
+                tests: [],
+                errorMessages: [errorMsg],
+                status: RunStatus.Error
+              });
+            }
           });
         } catch (error) {
           resolve({

--- a/packages/stryker-mocha-runner/src/StrykerMochaReporter.ts
+++ b/packages/stryker-mocha-runner/src/StrykerMochaReporter.ts
@@ -6,12 +6,15 @@ const log = log4js.getLogger('StrykerMochaReporter');
 
 export default class StrykerMochaReporter {
 
-  private runResult: RunResult;
+  public runResult: RunResult;
   private timer = new Timer();
   private passedCount = 0;
 
+  static CurrentInstance: StrykerMochaReporter | undefined;
+
   constructor(private runner: NodeJS.EventEmitter) {
     this.registerEvents();
+    StrykerMochaReporter.CurrentInstance = this;
   }
 
   private registerEvents() {
@@ -53,7 +56,6 @@ export default class StrykerMochaReporter {
 
     this.runner.on('end', () => {
       this.runResult.status = RunStatus.Complete;
-      (this.runner as any).runResult = this.runResult;
       log.debug(`Mocha test run completed: ${this.passedCount}/${this.runResult.tests.length} passed`);
     });
   }

--- a/packages/stryker-mocha-runner/test/integration/MochaTestRunnerSpec.ts
+++ b/packages/stryker-mocha-runner/test/integration/MochaTestRunnerSpec.ts
@@ -87,5 +87,28 @@ describe('MochaTestRunner', function () {
     }));
   });
 
-  let file = (filePath: string, mutated: boolean = true, included: boolean = true) => ({ path: path.resolve(filePath), mutated, included });
+  describe('when no tests are executed', () => {
+
+    beforeEach(() => {
+      const testRunnerOptions = {
+        files: [
+          file('./testResources/sampleProject/src/MyMath.js')],
+        strykerOptions: {},
+        port: 1234
+      };
+      sut = new MochaTestRunner(testRunnerOptions);
+    });
+
+    it('should report no completed tests', () =>
+      expect(sut.run()).to.eventually.satisfy((runResult: RunResult) => {
+        expect(countSucceeded(runResult)).to.be.eq(0, 'Succeeded tests did not match');
+        expect(countFailed(runResult)).to.be.eq(0, 'Failed tests did not match');
+        runResult.tests.forEach(t => expect(t.timeSpentMs).to.be.greaterThan(-1).and.to.be.lessThan(1000));
+        expect(runResult.status).to.be.eq(RunStatus.Complete, 'Test result did not match');
+        expect(runResult.coverage).to.not.be.ok;
+        return true;
+      }));
+  });
+
+  let file = (name: string, mutated: boolean = true, included: boolean = true) => ({ path: path.resolve(name), mutated, included });
 });

--- a/packages/stryker/src/Stryker.ts
+++ b/packages/stryker/src/Stryker.ts
@@ -71,7 +71,11 @@ export default class Stryker {
 
     const inputFiles = await new InputFileResolver(this.config.mutate, this.config.files).resolve();
     const { runResult, sandboxCoordinator } = await this.initialTestRun(inputFiles);
-    if (runResult && inputFiles && sandboxCoordinator) {
+
+     if (runResult.tests.length === 0) {
+      log.warn('No tests were executed. Stryker will exit prematurely. Please check your configuration.');
+      return [];
+    } else if (runResult && inputFiles && sandboxCoordinator) {
       const mutantResults = await this.generateAndRunMutations(inputFiles, runResult, sandboxCoordinator);
       const score = ScoreResultCalculator.calculate(mutantResults);
       this.reporter.onScoreCalculated(score);

--- a/packages/stryker/test/unit/StrykerSpec.ts
+++ b/packages/stryker/test/unit/StrykerSpec.ts
@@ -35,7 +35,7 @@ describe('Stryker', function () {
   let testFrameworkOrchestratorMock: Mock<TestFrameworkOrchestrator>;
   let configValidatorMock: Mock<ConfigValidator>;
   let sandboxCoordinatorMock: Mock<SandboxCoordinator>;
-  let configReaderMock: Mock<ConfigReader>; 
+  let configReaderMock: Mock<ConfigReader>;
   let pluginLoaderMock: Mock<PluginLoader>;
   let inputFiles: InputFile[];
   let determineExitCodeStub: sinon.SinonStub;
@@ -43,7 +43,7 @@ describe('Stryker', function () {
   let config: any;
   let mutants: any[];
   let reporter: Reporter;
-  
+
   beforeEach(() => {
     sandbox = sinon.sandbox.create();
     config = {};
@@ -286,6 +286,25 @@ describe('Stryker', function () {
                 it('should clean the stryker temp folder', () => expect(StrykerTempFolder.clean).to.have.been.called);
               });
             });
+          });
+        });
+
+        describe('with no tests executed', () => {
+          beforeEach(() => {
+            resolveInitialTestRun({
+              status: RunStatus.Complete,
+              tests: []
+            });
+          });
+
+          it('should log to have quit early', async () => {
+            await sut.runMutationTest();
+            expect(log.warn).to.have.been.calledWith('No tests were executed. Stryker will exit prematurely. Please check your configuration.');
+          });
+
+          it('should not have tested mutations', async () => {
+            await sut.runMutationTest();
+            expect(sandboxCoordinatorMock.runMutants).not.to.have.been.called;
           });
         });
       });

